File: src/utils/emi.js
export function generateEmiSchedule({ principal, annualRatePercent, tenureMonths, startMonthIndex = 0 }) {
  const r = (annualRatePercent / 12) / 100
  const n = tenureMonths
  const p = principal
  if (n <= 0) return []
  const emi = r === 0 ? p / n : (p * r * Math.pow(1 + r, n)) / (Math.pow(1 + r, n) - 1)
  let balance = p
  const rows = []
  for (let i = 1; i <= n; i++) {
    const interest = r * balance
    const principalComponent = Math.min(emi - interest, balance)
    balance = Math.max(0, balance - principalComponent)
    rows.push({
      installment: i,
      payment: emi,
      principal: principalComponent,
      interest,
      balance,
      monthIndex: startMonthIndex + i - 1,
    })
  }
  return rows
}

File: Backend/src/main/java/com/vortexaBackend/EmiWallah/Service/Emi.java
package com.vortexaBackend.EmiWallah.Service;

import com.vortexaBackend.EmiWallah.Model.EMISchedules;
import com.vortexaBackend.EmiWallah.Model.Loans;
//import com.vortexaBackend.EmiWallah.Repository.EMISchedulesRepository;
import com.vortexaBackend.EmiWallah.Repository.EmiRepo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Service
public class Emi {

    @Autowired
    private EmiRepo emiSchedulesRepository;

    // Generate EMI schedules for a loan
    public List<EMISchedules> generateEMISchedule(Loans loan) {
        List<EMISchedules> emiList = new ArrayList<>();

        double principal = loan.getPrincipalAmount();
        double annualRate = loan.getInterestRate();
        int months = loan.getTenureMonths();

        double monthlyRate = annualRate / 12 / 100;

        double emi = (principal * monthlyRate * Math.pow(1 + monthlyRate, months))
                / (Math.pow(1 + monthlyRate, months) - 1);

        double remainingPrincipal = principal;
        LocalDate dueDate = loan.getStartDate().plusMonths(1);

        for (int i = 1; i <= months; i++) {
            double interest = remainingPrincipal * monthlyRate;
            double principalComponent = emi - interest;
            remainingPrincipal -= principalComponent;

            EMISchedules emiSchedule = new EMISchedules();
            emiSchedule.setLoan(loan);
            emiSchedule.setDueDate(dueDate);
            emiSchedule.setEmiAmount(Math.round(emi * 100.0) / 100.0);
            emiSchedule.setPrincipalComponent(Math.round(principalComponent * 100.0) / 100.0);
            emiSchedule.setInterestComponent(Math.round(interest * 100.0) / 100.0);
            emiSchedule.setOutstandingBalance(Math.round(Math.max(remainingPrincipal, 0) * 100.0) / 100.0);
            emiSchedule.setStatus("Pending");

            emiList.add(emiSchedule);
            dueDate = dueDate.plusMonths(1);
        }

        // Optionally save all at once
        emiSchedulesRepository.saveAll(emiList);

        return emiList;
    }
}
